libs/ui/src/battery_low_alert.tsx

```diff
@@ -0,0 +1,94 @@
+import React, { useState } from 'react';
+import { format } from '@votingworks/utils';
+import { Modal } from './modal';
+import { useQueryChangeListener } from './hooks/use_change_listener';
+import { useSystemCallApi } from './system_call_api';
+import { Button } from './button';
+import { H1, P } from './typography';
+import { Icons } from './icons';
+
+interface BatteryWarning {
+  level: number;
+  canDismiss: boolean;
+}
+
+const BATTERY_WARNINGS: BatteryWarning[] = [
+  { level: 0.15, canDismiss: true },
+  { level: 0.1, canDismiss: true },
+  { level: 0.05, canDismiss: true },
+  { level: 0.01, canDismiss: false },
+];
+
+function roundLevel(n: number): number {
+  return Math.round(n * 100) / 100;
+}
+
+/**
+ * A helper for SessionTimeLimitTracker
+ */
+export function BatteryLowAlert(): JSX.Element | null {
+  const systemCallApi = useSystemCallApi();
+  const batteryInfoQuery = systemCallApi.getBatteryInfo.useQuery();
+  const batteryInfo = batteryInfoQuery.data;
+  const [currentWarning, setCurrentWarning] = useState<BatteryWarning>();
+
+  useQueryChangeListener(batteryInfoQuery, {
+    onChange: (newBatteryInfo, previousBatteryInfo) => {
+      if (!newBatteryInfo) return;
+
+      // if the battery's charging, dismiss any existing warnings
+      if (!newBatteryInfo.discharging) {
+        setCurrentWarning(undefined);
+        return;
+      }
+
+      // if the rounded battery level hasn't changed, don't do anything
+      // unless the battery just started discharging
+      const newLevel = roundLevel(newBatteryInfo.level);
+      if (
+        previousBatteryInfo &&
+        newLevel === roundLevel(previousBatteryInfo.level) &&
+        previousBatteryInfo.discharging
+      ) {
+        return;
+      }
+
+      // if there is an applicable warning, show it
+      const applicableWarning = BATTERY_WARNINGS.find(
+        (w) => w.level === newLevel
+      );
+      if (applicableWarning) {
+        setCurrentWarning(applicableWarning);
+      }
```

Comment:
The logic of this component is that it gives a warning when the battery is at specific levels, rather than ranges. It keeps it slightly simpler and means we manage less state. But it means that, for example, if you booted up your computer at 14% you wouldn't get a warning immediately, not until the battery drops to 10%. We also are making the assumption that battery decline is smooth and passes through all percentiles, rather than dropping and skipping levels. That is the behavior of my development machine battery, not sure if that's universal.

Reply:
Fair point. I realized this was how it worked when I was writing testing but wanted some other input. Fixing this up now.